### ISSUE TITLE
Vector iterator with custom typemap memory management causes crash

### ISSUE THREAD

**Issue by drivehappy** [swig/swig#1038]
I've built out a custom typemap to wrap a custom C++ String type. The interesting pieces of this typemap are:

```
// Helper function to cleanup memory allocated on the native side
%{
    void DeleteArrayMemory(void* m) {
        delete[] reinterpret_cast<char*>(m);
    }
%}
void DeleteArrayMemory(void* m);


%typemap(out) String %{
    // Retrieves a UTF16 2-byte buffer from String and allocates a char16_t*
    
    int bufSize = $1.size() * 2;
    char16_t* s = new char16_t[bufSize + 2];
    memcpy(s, $1.getBuffer(), bufSize);
    memset(((char*)s + bufSize), 0, 2);
    $result = (void*)s;
%}

%typemap(csout, excode=SWIGEXCODE) String {
    // Accepts the char16_t* allocated and deallocates the memory at GC
    
    var s = $imcall;
    string ret = global::System.Runtime.InteropServices.Marshal.PtrToStringUni(s);$excode;
    var hr = new System.Runtime.InteropServices.HandleRef(this, s);
    $imclassname.DeleteArrayMemory(hr);
    return ret;
}
```

I have a method on a C++ object that returns a std::vector<String> type:

```
class Bar {
    std::vector<String> foo() const;
}
```

And use the SWIG std_vector.i to expose the String:

```
%include <std_vector.i>

// snip, other includes for the above types

SWIG_STD_VECTOR_ENHANCED(String)
%template(String_Vector) std::vector<String>;
```

This compiles, however at runtime if the C# code accesses this foo() as an iterator, e.g.:

```
void test() {
    Bar b;
    foreach (string s in b.foo()) { // Crash here
                                    // Directly caused by the DeleteArrayMemory above
    }
}
```

I tracked this down to the following stack trace:
```
DeleteArrayMemory
[Managed to Native Transition]
String_Vector.getitem(int index)
String_Vector.this[int].get(int index)
String_Vector.String_VectorEnumerator.MoveNext()
test()
```

It appears that the getitem() call in std_vector.i is not performing the expected allocation of the String -> char16_t* (from the typemap(out)).
However, when the pointer is received on the managed side it assumes it has been allocated by my custom typemap and attempts to delete it.

Replacing the implementation in MoveNext():

`currentObject = collectionRef[currentIndex];`
    
To:

`currentObject = collectionRef.getitemcopy(currentIndex);`

Seems to resolve the problem. Obviously, there might be a performance concern here if we're always copying objects in other cases unnecessarily.

It seems like it's an issue in the std_vector implementation, as I'm unsure if there's anything I can do in my typemap to prevent this case.

**Comment by ojwb** [comment]
You seem to allocate the memory as an array of `char16_t` but delete it as an array of `char` - I'd make those types agree and see if that helps.

Also, the out typemap code seems confused about whether `bufSize` is the number of bytes or the number of `char16_t`, but by my reading I think it just allocates twice the needed space.

**Comment by drivehappy** [comment]
@ojwb: You're correct on both accounts, I originally was using a `char` type, so the pointer math changed. However, this did not resolve the issue.

After reviewing this again (as I ran into the same issue with std::map), it appears that the typemap I used for "const String&" (which I omitted in the original issue description) was incorrect:

```
%typemap(out) String %{ 
    ...
%}

%typemap(csout, excode=SWIGEXCODE) const String & {
    ...
}
```

Which should have been:
```
%typemap(out) const String & %{ 
    ...
%}

%typemap(csout, excode=SWIGEXCODE) const String & {
    ...
}
```

With this fix on my side everything looks like it works correctly now.
